lib/vocabularies/discriminator/index.ts

```diff
@@ -69,18 +69,46 @@ const def: CodeKeywordDefinition = {
           sch = resolveRef.call(it.self, it.schemaEnv.root, it.baseId, sch?.$ref)
           if (sch instanceof SchemaEnv) sch = sch.schema
         }
-        const propSch = sch?.properties?.[tagName]
-        if (typeof propSch != "object") {
+        let propSch = sch?.properties?.[tagName]
+        let hasSubSchRequired = false
+        if (!propSch && sch?.allOf) {
+          const {hasRequired, propertyObject} = mapDiscriminatorFromAllOf(propSch, sch)
+          hasSubSchRequired = hasRequired
+          propSch = propertyObject
+        }
+        if (!propSch || typeof propSch != "object") {
           throw new Error(
             `discriminator: oneOf subschemas (or referenced schemas) must have "properties/${tagName}"`
           )
         }
-        tagRequired = tagRequired && (topRequired || hasRequired(sch))
+        tagRequired = tagRequired && (topRequired || hasRequired(sch) || hasSubSchRequired)
         addMappings(propSch, i)
       }
       if (!tagRequired) throw new Error(`discriminator: "${tagName}" must be required`)
       return oneOfMapping
 
+      function mapDiscriminatorFromAllOf(
+        propSch: any,
+        sch: any
+      ): {hasRequired: boolean; propertyObject: any} {
+        let subSchObj: any = null
+        for (const subSch of sch.allOf) {
+          if (subSch?.properties) {
+            propSch = subSch.properties[tagName]
+            subSchObj = subSch
+          } else if (subSch?.$ref) {
+            subSchObj = resolveRef.call(it.self, it.schemaEnv.root, it.baseId, subSch.$ref)
+            if (subSchObj instanceof SchemaEnv) subSchObj = subSchObj.schema
+            propSch = subSchObj?.properties?.[tagName]
+          }
+          if (propSch) {
```

Comment:
is it possible to move the check out of `for`? The reason is that currently the first match is applied and it is better to return the last match. Therefore the last override has priority